src/net/socket.rs

```diff
@@ -0,0 +1,222 @@
+use crate::config::NetworkConfig;
+use crate::error::{NetworkError, NetworkErrorKind, NetworkResult};
+use crate::net::link_conditioner::LinkConditioner;
+use crate::net::{connection::ActiveConnections, events::SocketEvent};
+use crate::packet::Packet;
+use log::error;
+use mio::{Evented, Events, Poll, PollOpt, Ready, Token};
+use std::io;
+use std::mem;
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::{mpsc, Arc};
+
+const SOCKET: Token = Token(0);
+
+/// A reliable UDP socket implementation with configurable reliability and ordering guarantees.
+pub struct LaminarSocket {
+    socket: mio::net::UdpSocket,
+    config: Arc<NetworkConfig>,
+    connections: ActiveConnections,
+    recv_buffer: Vec<u8>,
+    link_conditioner: Option<LinkConditioner>,
+    event_sender: mpsc::Sender<SocketEvent>,
+    packet_receiver: mpsc::Receiver<Packet>,
+}
+
+impl LaminarSocket {
+    /// Binds to the socket and then sets up `ActiveConnections` to manage the "connections".
+    /// Because UDP connections are not persistent, we can only infer the status of the remote
+    /// endpoint by looking to see if they are still sending packets or not
+    pub fn bind<A: ToSocketAddrs>(
+        addresses: A,
+        config: NetworkConfig,
+    ) -> NetworkResult<(Self, mpsc::Sender<Packet>, mpsc::Receiver<SocketEvent>)> {
+        let socket = std::net::UdpSocket::bind(addresses)?;
+        let socket = mio::net::UdpSocket::from_socket(socket)?;
+        Ok(Self::new(socket, config))
+    }
+
+    /// Entry point to the run loop. This should run in a spawned thread since calls to `poll.poll`
+    /// are blocking.
+    pub fn start_polling(&mut self) -> NetworkResult<()> {
+        let poll = Poll::new()?;
+
+        poll.register(self, SOCKET, Ready::readable(), PollOpt::edge())?;
+
+        let mut events = Events::with_capacity(self.config.socket_event_buffer_size);
+        let events_ref = &mut events;
+        // Packet receiver MUST only be used in this method.
+        let packet_receiver = mem::replace(&mut self.packet_receiver, mpsc::channel().1);
+        // Nothing should break out of this loop!
+        loop {
```

Comment:
Is there a way to stop the polling or are we even supposed to stop the polling?

Reply:
That's a good question. I mean, if we stop polling we'll stop receiving packets but perhaps it's useful to have.